### PR TITLE
Fix pickling of exceptions

### LINKED ISSUE #4000: Exceptions are not pickle-able
### Problem you have encountered:

Flax exceptions can not be pickled. This makes it impossible to trace where/why errors are occurring when calling flax via ray.

### System information
```
Name: flax
Version: 0.8.3
Summary: Flax: A neural network library for JAX designed for flexibility
Home-page:
Author:
Author-email: Flax team <flax-dev@google.com>
License:
Location: /home/sander_cohere_com/miniconda3/envs/ct/lib/python3.10/site-packages
Requires: jax, msgpack, numpy, optax, orbax-checkpoint, PyYAML, rich, tensorstore, typing-extensions
Required-by: dm-haiku, fax
---
Name: jax
Version: 0.4.28
Summary: Differentiate, compile, and transform Numpy code.
Home-page: https://github.com/google/jax
Author: JAX team
Author-email: jax-dev@google.com
License: Apache-2.0
Location: /home/sander_cohere_com/miniconda3/envs/ct/lib/python3.10/site-packages
Requires: ml-dtypes, numpy, opt-einsum, scipy
Required-by: chex, fax, flax, optax, orbax-checkpoint, rax
---
Name: jaxlib
Version: 0.4.28
Summary: XLA library for JAX
Home-page: https://github.com/google/jax
Author: JAX team
Author-email: jax-dev@google.com
License: Apache-2.0
Location: /home/sander_cohere_com/miniconda3/envs/ct/lib/python3.10/site-packages
Requires: ml-dtypes, numpy, scipy
Required-by: chex, fax, optax, orbax-checkpoint, rax
```

### What you expected to happen:

Flax Exceptions to support pickle, cf [this guide](https://stackoverflow.com/questions/16244923/how-to-make-a-custom-exception-class-with-multiple-init-args-pickleable)

### Logs, error messages, etc:

In production code:
```
site-packages/ray/exceptions.py", line 49, in from_ray_exception
    return pickle.loads(ray_exception.serialized_exception)
TypeError: ScopeParamShapeError.__init__() missing 3 required positional arguments: 'scope_path', 'value_shape', and 'init_shape'
```

### Steps to reproduce:

```
import flax.linen as nn
from jax import random
from flax.linen.initializers import lecun_normal
from jax import lax
import pickle

class NoBiasDense(nn.Module):
  features: int = 8

  @nn.compact
  def __call__(self, x):
    kernel = self.param('kernel',
                        lecun_normal(),
                        (x.shape[-1], self.features))  # <--- Exception from flax docs example
    y = lax.dot_general(x, kernel,
                        (((x.ndim - 1,), (0,)), ((), ())))
    return y

variables = NoBiasDense().init(random.key(0), jnp.ones((5, 5, 1)))
try:
    _ = NoBiasDense().apply(variables, jnp.ones((5, 5)))
except Exception as e:
    str = pickle.dumps(e)
    obj = pickle.loads(str) # <--- pickle exception
```

### PR DESCRIPTION
**Pull request by sanderland**
# What does this PR do?

Fixes #4000

## Checklist
- [x] The documentation and docstrings adhere to the
      [documentation guidelines](https://github.com/google/flax/blob/main/docs/README.md#how-to-write-code-documentation).
- [x] This change includes necessary high-coverage tests.
      (No quality testing = no merge!)

### CODE REVIEW

File: tests/exceptions_pickle_test.py
Comment (IvyZX) [inline tests/exceptions_pickle_test.py:1]: Rename this to `pickle_test`?

File: tests/exceptions_pickle_test.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# limitations under the License.
+
+"""Tests for flax.errors."""
+
+from absl.testing import absltest
+from flax.errors import FlaxError, ScopeVariableNotFoundError
+import pickle
+
+class ErrorrsTest(absltest.TestCase):
+  def test_exception_can_be_picked(self):
```
Comment (IvyZX) [inline tests/exceptions_pickle_test.py:22]: nit: `test_exception_can_be_pickled`?

**Comment by sanderland** [comment]
Thanks @IvyZX, I've made the small fixes. We'd appreciate a release when the fix is merged, as it is currently affecting our ability to debug.

**Review by IvyZX — APPROVED** [review]
Thank you!

**Pull request merged**